src/ORM/FieldType/DBGenerated.php

```diff
@@ -0,0 +1,288 @@
+<?php
+
+namespace SilverStripe\ORM\FieldType;
+
+use InvalidArgumentException;
+use SilverStripe\Assets\Storage\DBFile;
+use SilverStripe\Core\Validation\ValidationResult;
+use SilverStripe\Forms\FormField;
+use SilverStripe\Model\ModelData;
+use SilverStripe\ORM\DB;
+use SilverStripe\ORM\Filters\SearchFilter;
+
+/**
+ * A column which gets generated based on some SQL statement.
+ * This can be stored (generated whenever columns in the generation statement are updated), or virtual (generated on the fly when accessed).
+ */
+class DBGenerated extends DBField
+{
+    public const string GENERATION_STORED = 'STORED';
+
+    public const string GENERATION_VIRTUAL = 'VIRTUAL';
+
+    /**
+     * DBField classes which must not be used as the underlying data representation for a generated column.
+     */
+    private static array $invalid_child_field_types = [
+        DBClassName::class,
+        DBClassNameVarchar::class,
+        DBComposite::class,
+        DBFile::class,
+        DBForeignKey::class,
+        DBGenerated::class,
+        DBPolymorphicForeignKey::class,
+        DBPrimaryKey::class,
+    ];
+
+    private DBField $childField;
+
+    private string $sqlExpression;
+
+    private string $generationType;
+
+    public function __construct(?string $name = null, string $fieldSpec = '', string $sqlExpression = '', string $type = DBGenerated::GENERATION_VIRTUAL)
```

Comment:
```suggestion
    public function __construct(
          ?string $name = null,
          string $fieldSpec = '',
          string $sqlExpression = '',
          string $generationType = DBGenerated::GENERATION_VIRTUAL
      )
```
Split over multi-lines because it was getting long

Main change is $type => $generationType so that the param matches how we're referring to it everywhere else

Reply:
This feels like it should be added to the [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/php_coding_conventions/) since you prefer it this way and I don't lol. That way at least there is an objectively right answer instead of me always doing it the way I prefer it and you always asking me to change it.

I'll make this change this time but please consider updating the coding conventions.

Reply:
Done

Reply:
Have made issue + pr https://github.com/silverstripe/developer-docs/issues/781

Reply:
Might be a good way to start using https://php.watch/versions/8.0/constructor-property-promotion